solr/core/src/java/org/apache/solr/core/CoreContainer.java

```diff
@@ -445,6 +447,17 @@ public CoreContainer(NodeConfig config, CoresLocator locator, boolean asyncSolrC
             cfg.getIndexSearcherExecutorThreads(), // thread count
             cfg.getIndexSearcherExecutorThreads(), // queue size
             new SolrNamedThreadFactory("searcherCollector"));
+    ((ExecutorUtil.MDCAwareThreadPoolExecutor) collectorExecutor).setRejectedExecutionHandler(new RejectedExecutionHandler() {
```

Comment:
I think that's an anti-pattern or broken and isn't what I meant in JIRA.  We could use a SynchronousQueue (with fairness) if we want to block for a thread -- probably what we should do.  FYI that queue is the default for `Executors.newCachedThreadPool()`.    The "caller runs" behavior I meant could be done via an ExecutorService delegate that catches RejectedException and simply runs the Runnable.

Reply:
> The "caller runs" behavior I meant could be done via an ExecutorService delegate that catches RejectedException and simply runs the Runnable.

I don't think  "caller runs" is a good solution. If there are multiple rejected tasks because all threads are busy, the caller thread can only run one task, what will happen to the other tasks? Meanwhile the already completed tasks may free up other threads. So, IMHO, the current code with a bigger queue count should be fine

Reply:
Have you considered SynchronousQueue instead of a big queue?  I'm not sure what value the queue has for this use-case.

Reply:
Actually, a SynchronousQueue is good enough too. 

So, the main thread will automatically wait for a pool thread to be available for executing the tasks